src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+                None,
+            ));
+        }
+
+        let value = parts.next().ok_or_else(|| {
+            Error::BadRemoteIndexFormat(
+                format!("Expected a value after '{}': '{}'", key, line),
+                None,
+            )
+        })?;
+
+        if parts.next().is_some() {
+            return Err(Error::BadRemoteIndexFormat(
+                format!("Too many values: '{}'", line),
+                None,
+            ));
+        }
+
+        Ok(value.to_string())
+    }
+
+    fn is_field_separator(ch: char) -> bool {
+        ch == ' ' || ch == '\t'
+    }
+}
+
+impl std::fmt::Display for RemoteIndex {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        if let Some(stem) = self.path().and_then(|p| p.file_stem()) {
+            write!(fmt, "{} ({})", stem.to_string_lossy(), self.url)?;
+        } else {
+            let stem = self.url.rsplit_once('/').unwrap().0;
+            write!(fmt, "{} ({})", stem, self.url)?;
+        }
+        Ok(())
+    }
+}
+
+/// Loads all remote index (.esi) files from the target directory.
+pub fn load_remotes(base_path: &Path) -> Result<Vec<RemoteIndex>, Error> {
+    let paths = fs::read_dir(base_path).map_err(Error::IOError)?;
+    paths
+        .filter_map(|e| e.ok())
+        .filter(|p| p.path().extension() == Some(OsStr::new(REMOTE_INDEX_EXTENSION)))
+        .map(|p| RemoteIndex::load(p.path()).reify(p.path().display()))
+        .collect()
+}
+
+/// Sends an HTTP GET request with the specified URL.
+///
+/// # Arguments
+/// * `url` - the URL to fetch
+/// * `timeout` - the timeout for the whole of the request
+/// * `size_limit` - the maximum size of the response body, before the
+///   connection is terminated
+/// * `if_modified_since` - Sets the value of the `If-Modified-Since` HTTP
+///   header
+fn open_remote_resource(
+    url: &Url,
+    timeout: Option<Duration>,
+    if_modified_since: Option<SystemTime>,
+) -> Result<Option<(usize, impl Read)>, Error> {
+    let mut agent_builder = AgentBuilder::new();
+    // Alternatively, we could have used Duration::MAX to indicate no timeout.
+    // Unfortunately, the ureq crashes at some unwrap() somewhere then MAX is
+    // provided.
+    if let Some(timeout) = timeout {
+        agent_builder = agent_builder.timeout(timeout);
+    }
+    let agent = agent_builder.build();
+
+    let mut request = agent.get(url.as_ref());
+    if let Some(if_modified_since) = if_modified_since {
+        request = request.set("If-Modified-Since", &format_http_date(if_modified_since));
+    }
+
+    let response = request.call().map_err(|e| Error::HttpError(e.into()))?;
+
+    let status = response.status();
+
+    let content_length = response
+        .header("Content-Length")
+        .unwrap_or("0")
+        .parse::<usize>()
+        .map_err(|_| {
+            Error::HttpError(
+                "The received Content-Length header contained an invalid value!".into(),
+            )
+        })?;
```

Comment:
If it fails to parse can we instead treat this in the same way as it being missing, given that the implementation must also cope with it being missing.

Reply:
If this fails to parse then the response is surely broken, right?

Reply:
Yes. My thinking here is that is a failure of the server rather than anything the user can fix. And it's probably not important enough for the user to need to do anything about it.